gpii/node_modules/userListeners/src/listeners.js

```diff
@@ -0,0 +1,176 @@
+/* User listeners
+ *
+ * Copyright 2017 Raising the Floor - International
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * The R&D leading to these results received funding from the
+ * Department of Education - Grant H421A150005 (GPII-APCP). However,
+ * these results do not necessarily represent the policy of the
+ * Department of Education, and you should not assume endorsement by the
+ * Federal Government.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+"use strict";
+
+var fluid = require("infusion"),
+    request = require("request");
+
+var gpii = fluid.registerNamespace("gpii");
+
+fluid.registerNamespace("gpii.userListeners");
+
+// The user listeners.
+fluid.defaults("gpii.userListeners", {
+    gradeNames: ["fluid.component", "fluid.contextAware"],
+    contextAwareness: {
+        platform: {
+            checks: {
+                windows: {
+                    contextValue: "{gpii.contexts.windows}",
+                    gradeNames: ["gpii.userListeners.windows"]
+                }
+            }
+        }
+    },
+    components: {
+        pcsc: {
+            type: "gpii.userListeners.pcsc"
+        },
+        usb: {
+            type: "gpii.userListeners.usb"
+        }
+    },
+    events: {
+        // The listeners are starting.
+        "onListenersStart": null,
+        // The listeners are stopping.
+        "onListenersStop": null
+    },
+    listeners: {
+        "onCreate.startListeners": "{that}.events.onListenersStart"
+    }
+});
+
+// A user listener.
+fluid.defaults("gpii.userListener", {
+    gradeNames: ["fluid.component"],
+    events: {
+        // A token has arrived.
+        "onTokenArrive": null,
+        // A token has removed.
+        "onTokenRemove": null,
+        // The listener has failed.
+        "onFail": null
+    },
+    invokers: {
+        // Start the listener, returning a Promise
+        startListener: "fluid.notImplemented",
+        // Stop the listener, returning a Promise
+        stopListener: "fluid.notImplemented",
+        // Parse a Buffer containing an NDEF message, returning a string.
+        decodeNDEF: {
+            funcName: "gpii.userListeners.decodeNDEF",
+            args: [
+                "{arguments}.0" // NDEF buffer
+            ]
+        },
+        // The listener has failed, and should be restarted.
+        failed: {
+            funcName: "gpii.userListeners.failed",
+            args: [
+                "{that}",
+                "{arguments}.0" // The error.
+            ]
+        }
+    },
+    members: {
+        // Set to true to call proximityTriggered, otherwise login/logout.
+        proximity: false,
+        listenerName: "@expand:fluid.identity({that}.typeName)",
```

Comment:
Why an expander rather than just a reference to {that}.typeName? Also, depending on typeName is a bit unstable since it may end up being replaced via options distributions, etc. Probably best to just write an ordinary member name that the various concrete listeners override

Reply:
I just wanted a sane default. Added a comment instead.